test/sslapitest.c

```diff
@@ -144,61 +163,116 @@ static int test_keylog_output(char *buffer, const SSL *ssl,
                                                          master_key_size);
             if (!master_key_size) {
                 printf("Error getting master key to compare.\n");
-                return -1;
+                return 0;
             }
             if (compare_hex_encoded_buffer(token, strlen(token),
                                            actual_master_key,
                                            master_key_size)) {
                 printf("Bad value for master key: %s\n", token);
-                return -1;
+                return 0;
+            }
+
+            master_secret_count++;
+        } else if ((strcmp(token, "CLIENT_HANDSHAKE_TRAFFIC_SECRET") == 0) ||
+                   (strcmp(token, "SERVER_HANDSHAKE_TRAFFIC_SECRET") == 0) ||
+                   (strcmp(token, "CLIENT_TRAFFIC_SECRET_0") == 0) ||
+                   (strcmp(token, "SERVER_TRAFFIC_SECRET_0") == 0)) {
+            /* TLSv1.3 secret. Tokens should be: 64 ASCII bytes of hex-encoded
+             * client random, and then the hex-encoded secret. In this case,
+             * we treat all of these secrets identically and then just
+             * distinguish between them when counting what we saw.
+             */
+            if (strcmp(token, "CLIENT_HANDSHAKE_TRAFFIC_SECRET") == 0)
+                client_handshake_secret_count++;
+            else if (strcmp(token, "SERVER_HANDSHAKE_TRAFFIC_SECRET") == 0)
+                server_handshake_secret_count++;
+            else if (strcmp(token, "CLIENT_TRAFFIC_SECRET_0") == 0)
+                client_application_secret_count++;
+            else if (strcmp(token, "SERVER_TRAFFIC_SECRET_0") == 0)
+                server_application_secret_count++;
+
+            client_random_size = SSL_get_client_random(ssl,
+                                                       actual_client_random,
+                                                       SSL3_RANDOM_SIZE);
+            if (client_random_size != SSL3_RANDOM_SIZE) {
+                printf("Unexpected short client random.\n");
+                return 0;
+            }
+
+            token = strtok(NULL, " \n");
+            if (!token) {
+                printf("Unexpected short client handshake secret log.\n");
+                return 0;
+            }
+            if (strlen(token) != 64) {
+                printf("Bad value for client random: %s\n", token);
+                return 0;
+            }
+            if (compare_hex_encoded_buffer(token, 64, actual_client_random,
+                                           client_random_size)) {
+                printf("Bad value for client random: %s\n", token);
+                return 0;
             }
 
-            saw_client_random = 1;
+            token = strtok(NULL, " \n");
+            if (!token) {
+                printf("Unexpectedly short master secret log.\n");
+                return 0;
+            }
+
+            /* TODO: Can I get this key out? */
```

Comment:
I'm thinking it might be useful to do at least the application traffic secrets bit - it has value as a test beyond the logging functionality. However, thinking about it more you would need to have access to the internal tls13_hkdf_expand() function which, wile doable, probably would mean moving the test to the tls13secretstest.c file. Perhaps a separate PR? For now just leave the TODO in, make it something like `TODO(TLS1.3): test that application traffic secrets are what we expect`